apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransaction.cs

```diff
@@ -0,0 +1,135 @@
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Cloud.Spanner.V1;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Spanner.Data
+{
+    internal sealed class RetriableTransaction
+    {
+        private readonly SpannerConnection _connection;
+        private readonly IClock _clock;
+        private readonly IScheduler _scheduler;
+        private readonly RetriableTransactionOptions _options;
+
+        internal RetriableTransaction(SpannerConnection connection, IClock clock, IScheduler scheduler, RetriableTransactionOptions options = null)
+        {
+            _connection = GaxPreconditions.CheckNotNull(connection, nameof(connection));
+            _clock = GaxPreconditions.CheckNotNull(clock, nameof(clock));
+            _scheduler = GaxPreconditions.CheckNotNull(scheduler, nameof(scheduler));
+            _options = options ?? RetriableTransactionOptions.CreateDefault();
+        }
+
+        public async Task<TResult> RunAsync<TResult>(Func<SpannerTransaction, Task<TResult>> asyncWork, CancellationToken cancellationToken = default)
+        {
+            GaxPreconditions.CheckNotNull(asyncWork, nameof(asyncWork));
+
+            // Session will be initialized and subsequently modified by CommitAttempt.
+            PooledSession session = null;
+            try
+            {
+                return await ExecuteWithRetryAsync(CommitAttempt, cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                session?.Dispose();
+            }
+
+            async Task<TResult> CommitAttempt()
+            {
+                return await ExecuteHelper.WithErrorTranslationAndProfiling(
+                    async () =>
+                    {
+                        SpannerTransaction transaction = null;
+
+                        try
+                        {
+                            session = await (session?.WithFreshTransactionOrNewAsync(cancellationToken) ?? _connection.AcquireReadWriteSessionAsync(cancellationToken)).ConfigureAwait(false);
+                            transaction = new SpannerTransaction(_connection, TransactionMode.ReadWrite, session, null);
+
+                            TResult result = await asyncWork(transaction).ConfigureAwait(false);
+                            await transaction.CommitAsync(cancellationToken).ConfigureAwait(false);
+
+                            return result;
+                        }
+                        catch when (transaction != null)
+                        {
+                            try
+                            {
+                                await transaction.RollbackAsync(cancellationToken).ConfigureAwait(false);
+                            }
+                            catch (Exception e)
+                            {
+                                // If the commit actually aborted, calling Rollback will fail.
```

Comment:
I don't suppose we can tell whether that's the case and avoid even trying the rollback?

Reply:
No, we really can't tell if it was the commit that failed or not, or we could but with uglier code than this I think. I did changed the comment though, because rollback will fail if the commit was attempted and failed, but not only on aborted but other possible failures.